src/web/wms/src/main/java/org/geoserver/wms/web/data/AbstractStylePage.java

```diff
@@ -430,7 +432,59 @@ public void setRawStyle(Reader in) throws IOException {
         editor.setModelObject(rawStyle);
         in.close();
     }
-    
+
+    /**
+     * Check for an original CSS version of the style created by the old CSS extension (pre-pluggable styles). If a CSS style is found, recover it if
+     * the derived SLD has not subsequently been manually edited.
+     * 
+     * The recovery is accomplished by updating the catalog to point to the original CSS file, and changing the style's format to "css".
+     * 
+     * @param si The {@link StyleInfo} for which to check for and potentially recover a CSS version.
+     */
+    protected void recoverCssStyle(StyleInfo si) {
+        if (si == null) {
+            return;
+        }
+
+        // Only try to repair missing CSS files if a CSS style handler is registered.
+        try {
+            Styles.handler("css");
+        } catch (Exception e) {
+            return;
+        }
+
+        // Use this tolerance to prevent erasing an SLD that was manually edited after being
+        // generated from a CSS file. (Generated SLDs will always be newer than the CSS).
+        long favorSLDIfNewerByMS = 600000;
+
+        // The problem only exists for styles that were not explicitly set as "css".
+        if (!"css".equalsIgnoreCase(si.getFormat())) {
+            String filename = si.getFilename();
+            String filenameCss = filename.substring(0, filename.lastIndexOf('.')) + ".css";
+            GeoServerDataDirectory g = new GeoServerDataDirectory(getCatalog().getResourceLoader());
+            Resource cssResource = g.get(si, filenameCss);
+
+            if (!cssResource.getType().equals(Resource.Type.UNDEFINED)) {
+                // If there is an existing CSS file with the style's name, check if it should be recovered.
+                Resource sldResource = g.get(si, filename);
+                long sldNewerByMs = sldResource.lastmodified() - cssResource.lastmodified();
+
+                if (sldNewerByMs > favorSLDIfNewerByMS) {
+                    LOGGER.log(Level.WARNING,
+                            "A CSS version of this style has been recovered (" + filenameCss
```

Comment:
Rather than "this style", refer to the style by the StyleInfo name, since there will be no way of determining the current style when looking back at the logs.

e.g.
"A CSS version of the style OldCssStyle has been recovered ..."

If the style is in a workspace, the name should be prefaced by the workspace (<ws>:<name>) to ensure a unique name.